### DISCUSSION TITLE
Is there any way to wait until a store is persisted after a set call?

### DISCUSSION THREAD

**Discussion by shinobu-uwu** [pmndrs/zustand#2768] [Q&A]
Hello, I have a store that is being used both by a service worker and a web page. I'm facing a few synchronization issues, and what I did to fix them was calling `store.persist.rehydrate()` on my page after the service worker operation completed, however the values are still `undefined` after the rehydrate call, which is async btw. 

My guess is that it's being rehydrated with the wrong values because the service worker store didn't finish persisting the new values. So I decided to add a `sleep` call to the service worker call before returning and it started working, however this is not good and I'd like to know if there is a way to wait for the changes to be persisted.

**Comment by dai-shi** [comment]
Thanks for reporting.
I think it's a valid concern. Currently, we throw away the promise returned by `setItem` of the storage. It's not too hard to return a possible promise in JS, but TS typing will be extremely hard. I wonder if anyone wants to try.
In the meantime, the workaround will be to create a custom storage with which you can control more like getting a promise. Let us know if you need help.
